### PR TITLE
Add TempTableDefinitions and TableVarMapping

### PR DESCRIPTION
**Pull request by davidtme**
See #277 and https://github.com/fsprojects/FSharp.Data.SqlClient/issues/263#issuecomment-310667806

This allows you to add temp tables via the TempTableDefinitions parameter and table var types via the TableVarMapping parameter

### CODE REVIEW

**Comment by smoothdeveloper** [comment]
The features of this PR are interesting, I'd like to see those two changes:

* trim the changes not related to the PR (project files, fsharp.core, config file changes)
* add comments describing the implementation details
* possibly, add few unit tests to cover the string parsing logic of the implementation details, will help if we encounter bugs / corner cases in the future

Does the TVP type generated for the command unifies with the ones exposed in SqlProgrammabilityProvider's user defined types?

**Comment by vasily-kirichenko** [comment]
@davidtme I agree with @smoothdeveloper, could you please remove all unrelated changes, after which we could merge it? Thanks!

**Comment by davidtme** [comment]
I'll let #291 get in first then I'll go about making the changes suggested.

**Comment by vasily-kirichenko** [comment]
@davidtme OK. However, I'm stuck with #291, so no ETA.

**Comment by vasily-kirichenko** [comment]
@davidtme I'm not sure I'll finish the netstandard branch anytime soon (if ever), I think we should merge this PR. So, let's make the requested changes and we'll merge it.

**Comment by davidtme** [comment]
@smoothdeveloper The type(s) create to load data in to the temp tables live inside the command so are access like:

``` F#
use cmd = new TempTable(conn)
cmd.LoadTempTables([ TempTable.Temp(Id = 1, Name = Some "monkey") ])
```

As for TVP, no types are create by my code just the _design time_ sql is changed:

``` SQL
SELECT myId, myName from @input
```

Becomes something like (off the top of my head):

``` SQL
DECLARE @SQLCOMMANDPROVIDER_input dbo.MyTableType = @input
SELECT myId, myName from @SQLCOMMANDPROVIDER_input
```

**Comment by davidtme** [comment]
I'm not sure how get travis-ci to build the project, maybe it's a mono/old f# core thing as it complaining about

```
The module/namespace 'System.Collections.Generic' from compilation unit 'mscorlib' did not contain the namespace, module or type 'IReadOnlyCollection`1'
```

but it builds on windows:
```
FSharp.Data.TVPTests.UsingTVPInQuery [SKIP]
   Fails at runtime :(

FSharp.Data.TVPTests.InputIsEnumeratedExactlyOnce [SKIP]
   Flucky

FSharp.Data.TypeProviderTest.ConcurrentReaders [SKIP]
   Thread safe execution is not supported yet

FSharp.Data.TypeProviderTest.CommandTimeout [SKIP]
   Don't execute for usual runs. Too slow.

144 total, 0 failed, 4 skipped, took 3.531 seconds
Finished Target: RunTests
Starting Target: All (==> RunTests)
Finished Target: All

---------------------------------------------------------------------
Build Time Report
---------------------------------------------------------------------
Target            Duration
------            --------
Clean             00:00:00.0390081
RestorePackages   00:00:03.1203044
AssemblyInfo      00:00:00.0088544
Build             00:00:12.4519448
DeployTestDB      00:00:00.2427771
BuildTests        00:00:09.0156047
RunTests          00:00:04.2975552
All               00:00:00.0000459
Total:            00:00:29.2397542
Status:           Ok
---------------------------------------------------------------------
```

**Comment by vasily-kirichenko** [comment]
> I'm not sure how get travis-ci to build the project

Pin FAKE to version ~> 4.

**Comment by vasily-kirichenko** [comment]
OK, it does not help. The good news, I remembered the right way: reference FSharp.Core via NuGet. The version should be 4.0.0.1 I believe. 

@smoothdeveloper what version should we reference?

**Comment by vasily-kirichenko** [comment]
@davidtme But we must test in VS 2015 after such change.

**Comment by davidtme** [comment]
@vasily-kirichenko I did originally have 4.0.0.1 in there (a85d2b867d3fcd0a66efd00ac0d04a4df1570f73) but I removed it as part of the "clean up" and was hoped merging from master would sort it out... never mind it's a fine line between fixing to much and creating clutter in a PR so it's all good.  Sadly (not really) I no longer have VS2015 so I can't test that part.

**Comment by davidtme** [comment]
I've moved the build on a but

```
2) FS0039: /home/travis/build/fsprojects/FSharp.Data.SqlClient/src/SqlClient/ISqlCommand.fs(190,179): The type 'SqlDataRecord' is not defined in 'Microsoft.SqlServer.Server'.
```

Looks like this was in issue in the passed #49 so maybe pinning to 4.0.0.1 has resurrected this some how?

**Comment by vasily-kirichenko** [comment]
@davidtme I've built it from master branch and looked at FSharp.Core.dll in bin folder:

image

Try 4.1.18

**Comment by vasily-kirichenko** [comment]
No, it does not work. It seems the only way to make it compilable on Mac is upgrading to .net 4.5. Do we want to do it right now?

I think it's ok to have green build on windows for now.

**Comment by vasily-kirichenko** [comment]
I've built nuget package and testing in VS 2017:

image

**Comment by vasily-kirichenko** [comment]
OK, I have the same error with package built from master as well o_O

File: src/SqlClient/SqlCommandProvider.fs [hunk 1]
```diff
@@ -109,41 +109,51 @@ type SqlCommandProvider(config : TypeProviderConfig) as this =
         conn.CheckVersion()
         conn.LoadDataTypesMap()
 
-        let designTimeSqlStatement, tempTableTypes = 
-            let sql, types = DesignTime.SubstituteTempTables(conn, sqlStatement, tempTableDefinitions)
-            DesignTime.SubstituteTableVar(sql, tableVarMapping), types
+        let connectionId = Guid.NewGuid().ToString().Substring(0, 8)
```
Comment (smoothdeveloper) [inline src/SqlClient/SqlCommandProvider.fs]: Can we have explanation on why we need this? it seems to be used for the provided type name of the table, maybe better to pick something deterministic, or to remove it from here and internal only to the place it is used.

What would be the problem to just use the whole connection string instead?

Comment (smoothdeveloper) [reply src/SqlClient/SqlCommandProvider.fs]: or maybe the GetHashCode on the connection string, this way we can probably avoid passing a parameter and keep it as implementation detail of the substitution code.

Comment (davidtme) [reply src/SqlClient/SqlCommandProvider.fs]: In some cases parallel msbuild or msbuild and visual studio can both be building the type provider at the same time so you end up with both trying to create, use and drop the same temp table which creates build errors.  Adding a Guid in there stops them from fighting.

**Comment by davidtme** [comment]
Build is green :D I'm sure there is a cleaner way but could be a quick work around.

**Comment by smoothdeveloper** [comment]
@davidtme I think some of your changes for Mono are good but require more testing to verify we don't break existing runtime environments.

 I'm working on changes to bring up paket in master, I hope to submit a working PR this weekend, this should make dealing with fsharp.core and binding redirects easier.

**Comment by vasily-kirichenko** [comment]
I've merged master into your branch and opened new PR https://github.com/fsprojects/FSharp.Data.SqlClient/pull/302
